### DISCUSSION TITLE
INA226 not working on ESP8266 or ESP32

### DISCUSSION THREAD

**Discussion by VASTechAD** [arendst/Tasmota#18041] [Q&A] [answered]
### PROBLEM DESCRIPTION
I attach INA226 to I2c bus and tasmota detect module as INA219. I self-compiled a firmware with I2cDriver35, disabled from console I2cDriver14 for INA219, activated driver for INA226 with I2cDriver35 1. I try on both esp8266 and esp32.

### REQUESTED INFORMATION
_Make sure your have performed every step and checked the applicable boxes before submitting your issue. Thank you!_

- [X] Read the [Contributing Guide and Policy](https://github.com/arendst/Tasmota/blob/development/CONTRIBUTING.md) and [the Code of Conduct](https://github.com/arendst/Tasmota/blob/development/CODE_OF_CONDUCT.md)
- [X] Searched the problem in [issues](https://github.com/arendst/Tasmota/issues)
- [X] Searched the problem in [discussions](https://github.com/arendst/Tasmota/discussions)
- [X] Searched the problem in the [docs](https://tasmota.github.io/docs/FAQ)
- [ ] Searched the problem in the [chat](https://discord.gg/Ks2Kzd4)
- [X] Device used (e.g., Sonoff Basic): ESP-01 and ESP32
- [ ] Tasmota binary firmware version number used: 12.4.0.1
  - [ ] Pre-compiled
  - [x] Self-compiled
- [ ] Flashing tools used: WebFlash
- [X] Provide the output of command: `Backlog Template; Module; GPIO 255`:
```lua
  Configuration output here:
CMD: Backlog Template
RSL: RESULT = {"NAME":"ESP32-DevKit","GPIO":[1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,0,1,1,1,0,1,1,1,0,0,0,0,1,1,1,1,1,0,0,1],"FLAG":0,"BASE":1}
CMD: Module
RSL: RESULT = {"Module":{"1":"ESP32-DevKit"}}
CMD: GPIO 255`
RSL: RESULT = {"GPIO0":{"0":"None"},"GPIO1":{"0":"None"},"GPIO2":{"608":"I2C SCL1"},"GPIO3":{"0":"None"},"GPIO4":{"640":"I2C SDA1"},"GPIO5":{"0":"None"},"GPIO6":{"0":"None"},"GPIO7":{"0":"None"},"GPIO8":{"0":"None"},"GPIO9":{"0":"None"},"GPIO10":{"0":"None"},"GPIO11":{"0":"None"},"GPIO12":{"0":"None"},"GPIO13":{"0":"None"},"GPIO14":{"0":"None"},"GPIO15":{"0":"None"},"GPIO16":{"0":"None"},"GPIO17":{"0":"None"},"GPIO18":{"0":"None"},"GPIO19":{"0":"None"},"GPIO20":{"0":"None"},"GPIO21":{"0":"None"},"GPIO22":{"0":"None"},"GPIO23":{"0":"None"},"GPIO24":{"0":"None"},"GPIO25":{"0":"None"},"GPIO26":{"0":"None"},"GPIO27":{"0":"None"},"GPIO32":{"0":"None"},"GPIO33":{"0":"None"},"GPIO34":{"0":"None"},"GPIO35":{"0":"None"},"GPIO36":{"0":"None"},"GPIO37":{"0":"None"},"GPIO38":{"0":"None"},"GPIO39":{"0":"None"}}

```
- [ ] If using rules, provide the output of this command: `Backlog Rule1; Rule2; Rule3`:
```lua
  Rules output here:

```
- [x] Provide the output of this command: `Status 0`:
```lua
  STATUS 0 output here:
22:48:27.868 CMD: Status 0
22:48:27.872 RSL: STATUS = {"Status":{"Module":1,"DeviceName":"Tasmota","FriendlyName":["Tasmota"],"Topic":"tasmota_804EB0","ButtonTopic":"0","Power":0,"PowerOnState":3,"LedState":1,"LedMask":"FFFF","SaveData":1,"SaveState":1,"SwitchTopic":"0","SwitchMode":[0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0],"ButtonRetain":0,"SwitchRetain":0,"SensorRetain":0,"PowerRetain":0,"InfoRetain":0,"StateRetain":0,"StatusRetain":0}}
22:48:27.882 RSL: STATUS1 = {"StatusPRM":{"Baudrate":115200,"SerialConfig":"8N1","GroupTopic":"tasmotas","OtaUrl":"http://ota.tasmota.com/tasmota32/release/tasmota32.bin","RestartReason":"Software reset CPU","Uptime":"0T00:19:26","StartupUTC":"2023-02-23T21:29:01","Sleep":50,"CfgHolder":4617,"BootCount":14,"BCResetTime":"2023-02-23T22:16:14","SaveCount":35}}
22:48:27.891 RSL: STATUS2 = {"StatusFWR":{"Version":"12.4.0.1(tasmota32)","BuildDateTime":"2023-02-23T22:19:11","Core":"2_0_7","SDK":"4.4.4.20230218","CpuFrequency":160,"Hardware":"ESP32-D0WDQ6 rev.1","CR":"403/699"}}
22:48:27.895 RSL: STATUS3 = {"StatusLOG":{"SerialLog":2,"WebLog":2,"MqttLog":0,"SysLog":0,"LogHost":"","LogPort":514,"SSId":["Vasile-NoT",""],"TelePeriod":300,"Resolution":"558180C0","SetOption":["00008009","2805C80001000600003C5A0A192800000000","00000080","00006000","00004000","00000000"]}}
22:48:27.910 RSL: STATUS4 = {"StatusMEM":{"ProgramSize":1487,"Free":560,"Heap":153,"StackLowMark":3,"PsrMax":0,"PsrFree":0,"ProgramFlashSize":4096,"FlashSize":4096,"FlashChipId":"164020","FlashFrequency":80,"FlashMode":"DIO","Features":["00000809","9F9AD7CF","0015A001","B7F7BFCF","05DA9BC0","E0360DD7","480840D2","20200000","D43C482D","000080B1"],"Drivers":"1,2,3,4,5,7,8,9,10,11,12,14,16,17,20,21,24,26,27,29,34,35,38,50,52,59,60,62,63,66,82,86,87,88","Sensors":"1,2,3,5,6,7,8,9,10,11,12,13,14,15,17,18,19,20,21,22,26,31,34,37,39,40,42,43,45,51,52,54,55,56,58,59,64,66,67,74,85,92,95,98,103,127"}}
22:48:27.925 RSL: STATUS5 = {"StatusNET":{"Hostname":"tasmota-804EB0-3760","IPAddress":"10.10.10.37","Gateway":"10.10.10.1","Subnetmask":"255.255.255.0","DNSServer1":"10.10.10.1","DNSServer2":"0.0.0.0","Mac":"78:21:84:80:4E:B0","IP6Global":"","IP6Local":"fe80::7a21:84ff:fe80:4eb0","Ethernet":{"Hostname":"","IPAddress":"0.0.0.0","Gateway":"0.0.0.0","Subnetmask":"0.0.0.0","DNSServer1":"10.10.10.1","DNSServer2":"0.0.0.0","Mac":"00:00:00:00:00:00","IP6Global":"","IP6Local":""},"Webserver":2,"HTTP_API":1,"WifiConfig":4,"WifiPower":17.0}}
22:48:27.933 RSL: STATUS6 = {"StatusMQT":{"MqttHost":"","MqttPort":1883,"MqttClientMask":"DVES_%06X","MqttClient":"DVES_804EB0","MqttUser":"DVES_USER","MqttCount":0,"MAX_PACKET_SIZE":1200,"KEEPALIVE":30,"SOCKET_TIMEOUT":4}}
22:48:27.938 RSL: STATUS7 = {"StatusTIM":{"UTC":"2023-02-23T21:48:27","Local":"2023-02-23T22:48:27","StartDST":"2023-03-26T02:00:00","EndDST":"2023-10-29T03:00:00","Timezone":"+01:00","Sunrise":"07:44","Sunset":"18:22"}}
22:48:27.947 RSL: STATUS10 = {"StatusSNS":{"Time":"2023-02-23T22:48:27"}}
22:48:27.954 RSL: STATUS11 = {"StatusSTS":{"Time":"2023-02-23T22:48:27","Uptime":"0T00:19:26","UptimeSec":1166,"Heap":156,"SleepMode":"Dynamic","Sleep":50,"LoadAvg":19,"MqttCount":0,"Berry":{"HeapUsed":3,"Objects":43},"Wifi":{"AP":1,"SSId":"Vasile-NoT","BSSId":"7A:45:58:AF:F9:DD","Channel":6,"Mode":"11n","RSSI":100,"Signal":-49,"LinkCount":1,"Downtime":"0T00:00:02"}}}

```
- [ ] Set `weblog` to 4 and then, when you experience your issue, provide the output of the Console log:
```lua
  Console output here:


```

### TO REPRODUCE
_Plug INA226 device on I2c bus and see results_

### EXPECTED BEHAVIOUR
_To show correct device in main window, and in MQTT ._

### SCREENSHOTS
_PS Picture attached are from ESP8266 but issue is the same in ESP32_
_I2cDriver14 detects module INA 226 as INA219_
_![image](https://user-images.githubusercontent.com/57189088/221038901-759e259a-8f5f-4a74-8e1a-3b911f30b7ca.png)_
_I2cDriver35 ... nothing_
_![image](https://user-images.githubusercontent.com/57189088/221040131-77c03c2c-013d-4eca-a407-e7f06a0a00cf.png)_

### ADDITIONAL CONTEXT
_Add any other context about the problem here._

**(Please, remember to close the issue when the problem has been addressed)**

**Comment by barbudor** [comment]
Your build includes multiple drivers that use the same address as INA219 (i2cdriver14) and INA226 (i2cdriver35) such as HTU (i2cdriver9)
In order to use a INA226, you MUST disable drivers that you don't need and that can conflict
```
backlog i2cdriver14 0; i2cdriver15 0; i2cdriver35 1
```

**Comment by VASTechAD** [comment]
I`ve compiled a version for ESP8266 with i2cdriver35 only.

```lua
09:23:10.324 CMD: i2cdriver
09:23:10.330 MQT: stat/tasmota_24F95C/RESULT = {"I2CDriver":"35"}
```
```lua
09:26:33.277 CMD: i2cscan
09:26:33.299 MQT: stat/tasmota_24F95C/RESULT = {"I2CScan":"Device(s) found at 0x40"}
```
Same result:
![image](https://user-images.githubusercontent.com/57189088/221129608-264fd710-577a-4328-ae0d-73e0fa88b095.png)

Here a picture with device:
![image](https://user-images.githubusercontent.com/57189088/221130779-5ede093d-b92b-4265-96be-136ae6a9ac9c.png)

**Comment by Jason2866** [comment]
Please post your `user_config_override.h`.

**Comment by VASTechAD** [comment]
@Jason2866 
user_config_override.h is default, no modification there.

**Comment by sfromis** [comment]
Since you have only `{"I2CDriver":"35"}` you did modify your build to get to that, the key is not if it specifically was done in `user_config_override.h` (which is generally the preferred place for such) or not.

**Reply by VASTechAD** [reply]
I put `#define USE_INA226` in user_config_override.h aswell No changes!

**Comment by Jason2866** [comment]
Sorry, without providing relevant Infos we asked for, to help you with YOUR problem nobody can help.

**Reply by VASTechAD** [reply]
> Sorry, without providing relevant Infos we asked for, to help you with YOUR problem nobody can help.

`"/*
  user_config_override.h - user configuration overrides my_user_config.h for Tasmota

  Copyright (C) 2021  Theo Arends

  This program is free software: you can redistribute it and/or modify
  it under the terms of the GNU General Public License as published by
  the Free Software Foundation, either version 3 of the License, or
  (at your option) any later version.

  This program is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  GNU General Public License for more details.

  You should have received a copy of the GNU General Public License
  along with this program.  If not, see <http://www.gnu.org/licenses/>.
*/

#ifndef _USER_CONFIG_OVERRIDE_H_
#define _USER_CONFIG_OVERRIDE_H_

/*****************************************************************************************************\
 * USAGE:
 *   To modify the stock configuration without changing the my_user_config.h file:
 *   (1) copy this file to "user_config_override.h" (It will be ignored by Git)
 *   (2) define your own settings below
 *
 ******************************************************************************************************
 * ATTENTION:
 *   - Changes to SECTION1 PARAMETER defines will only override flash settings if you change define CFG_HOLDER.
 *   - Expect compiler warnings when no ifdef/undef/endif sequence is used.
 *   - You still need to update my_user_config.h for major define USE_MQTT_TLS.
 *   - All parameters can be persistent changed online using commands via MQTT, WebConsole or Serial.
\*****************************************************************************************************/

/*
Examples :

// -- Master parameter control --------------------
#undef  CFG_HOLDER
#define CFG_HOLDER        4617                   // [Reset 1] Change this value to load SECTION1 configuration parameters to flash

// -- Setup your own Wifi settings  ---------------
#undef  STA_SSID1
#define STA_SSID1         "YourSSID"             // [Ssid1] Wifi SSID

#undef  STA_PASS1
#define STA_PASS1         "YourWifiPassword"     // [Password1] Wifi password

// -- Setup your own MQTT settings  ---------------
#undef  MQTT_HOST
#define MQTT_HOST         "your-mqtt-server.com" // [MqttHost]

#undef  MQTT_PORT
#define MQTT_PORT         1883                   // [MqttPort] MQTT port (10123 on CloudMQTT)

#undef  MQTT_USER
#define MQTT_USER         "YourMqttUser"         // [MqttUser] Optional user

#undef  MQTT_PASS
#define MQTT_PASS         "YourMqttPass"         // [MqttPassword] Optional password

// You might even pass some parameters from the command line ----------------------------
// Ie:  export PLATFORMIO_BUILD_FLAGS='-DUSE_CONFIG_OVERRIDE -DMY_IP="192.168.1.99" -DMY_GW="192.168.1.1" -DMY_DNS="192.168.1.1"'

#ifdef MY_IP
#undef  WIFI_IP_ADDRESS
#define WIFI_IP_ADDRESS     MY_IP                // Set to 0.0.0.0 for using DHCP or enter a static IP address
#endif

#ifdef MY_GW
#undef  WIFI_GATEWAY
#define WIFI_GATEWAY        MY_GW                // if not using DHCP set Gateway IP address
#endif

#ifdef MY_DNS
#undef  WIFI_DNS
#define WIFI_DNS            MY_DNS               // If not using DHCP set DNS IP address (might be equal to WIFI_GATEWAY)
#endif

#ifdef MY_DNS2
#undef  WIFI_DNS2
#define WIFI_DNS2           MY_DNS2              // If not using DHCP set DNS IP address (might be equal to WIFI_GATEWAY)
#endif

// !!! Remember that your changes GOES AT THE BOTTOM OF THIS FILE right before the last #endif !!!
*/



#endif  // _USER_CONFIG_OVERRIDE_H_
`

**Reply by sfromis** [reply]
The "real" question was not specifically the one file `user_config_override.h` (which **is** the recommended file to modify), but about what you modified at all to get to your build.

**Reply by VASTechAD** [reply]
I only modified my_user_config.h
![image](https://user-images.githubusercontent.com/57189088/221170296-d187a610-4ca1-4396-943b-4450c9b42131.png)

**Reply by VASTechAD** [reply]
> Sorry, without providing relevant Infos we asked for, to help you with YOUR problem nobody can help.

Sorry! I not quite understood what I was do wrong, the issue is still there, I see that you close this issue and moved to discussion section. 
Please can you move it back to issue section?  Thank you!

**Reply by Jason2866** [reply]
No, as long required Infos are missing this is correctly placed in discussions.

**Reply by VASTechAD** [reply]
I`m sorry I do not understand what info you need or what i post wrong. can you be more precise? I`m not to good in this I try to do my best here. Thank you for your patience!

**Comment by VASTechAD** [comment]
[Jason2866](https://github.com/Jason2866)
I say that user_config_override.h was not altered. No modification there.

**Comment by Jason2866** [comment]
Still we don't know all the changes you did, posting a screenshots does not help.
We see just the adding of one define.  By doing this you would have additional enabled the ina226 driver.
Did you do "undefines"? Sorry you make everyone hard or better impossible to help.

**Comment by barbudor** [comment, marked as answer]
Have you read about [Sensor54](https://tasmota.github.io/docs/Commands/#sensor54) command ?

INA226 will not appear in the WebGUI until it first has been confirgured properly with Sensor54 command

**Reply by VASTechAD** [reply]
> Have you read about [Sensor54](https://tasmota.github.io/docs/Commands/#sensor54) command ?
> 
> INA226 will not appear in the WebGUI until it first has been confirgured properly with Sensor54 command



Thank you very much! @barbudor you are the man! That`s the solution. Thank you all!
 
![image](https://user-images.githubusercontent.com/57189088/221183240-89846d66-ef75-4b82-a8cf-24b330231e87.png)

**Reply by Ghvinerias** [reply]
@VASTechAD I am facing the same problem, can you provide Sensor54 commands that you used, I am using the same module and can not get it to work

**Reply by VASTechAD** [reply]
Hi! Just follow the instruction for Sensor54 command for the channel 1. Start with Sensor54 10
https://tasmota.github.io/docs/Commands/#sensor54
BR!

**Reply by Ghvinerias** [reply]
When I execute Sensor54 1 it reports it did not find any INA226 Devices, but when i execute i2cscan i can see that there is a i2c device with 0x40 address on the bus. As I figured out from your other comments you have added "#define USE_INA226" in user_config_override.h and compiled bin. Have you done anything else?

**Reply by VASTechAD** [reply]
Use the firmware 12.4.0.1 like in the above picture to see if is the same. ;)

**Reply by pkkrusty** [reply]
You need to define the resistance and shunt value with `Sensor54 11 .002` and `Sensor54 12 15`, then restart. It will then show correctly as INA226.

**Reply by barbudor** [reply]
The values are very dependent on the actual hardware
0.002 ohms is not the shunt which is by default on most Ali modules.

**Reply by Ghvinerias** [reply]
Thank All of you for your help, finally got it working, you guys rock!

**Comment by csh-de** [comment]
exactly what i need, would it be possible for you to send me the bin? I have no idea how to self-compile it.

**Reply by sfromis** [reply]
Doing your own compile is not hard. See https://tasmota.github.io/docs/Compile-your-build/
Using TasmoCompiler is probably the simplest.

**Comment by csh-de** [comment]
now it works but unfortunately the voltage is displayed without decimal places. Where can you adjust this?

![Screenshot_20230417_213646_Chrome](https://user-images.githubusercontent.com/130899114/232592949-7fa4777d-33e0-41c2-bf5b-1450a46ab10d.jpg)

**Reply by csh-de** [reply]
VoltRes 2

![Screenshot_20230417_215914_Chrome](https://user-images.githubusercontent.com/130899114/232597961-a497e6ed-fdd4-4e68-a761-b32cf60e40a3.jpg)

**Comment by tasmotari** [comment]
I have the same INA226 with R002 shunt, 
but I get incorrect voltage: 
13.06V battery = 5.22V
1.26V battery = 0.5V

How to fix it?

**Reply by barbudor** [reply]
Did you read the thread above as well as the documentation about Sensor54 ?

**Reply by tasmotari** [reply]
Yes, I set all parameters:
Sensor54 11 .002 (resistance)
Sensor54 12 20 (current)
Sensor54 2 (save & restart)

**Reply by barbudor** [reply]
If you get incorrect voltage, it most likely that you have incorrect cabling
Sensor54 commands only applies to current and shunt resistor
Correctly connected the woltage should be ok

Elaborate on your cabling

**Reply by tasmotari** [reply]
The cabling is so simple, there are only 2 connectors (+/-) for measuring voltage on sensor. 
The same cable use for multimeter and the value is correct. No problem with cabling.

I clearly see the difference between actual value and tasmota output in 2.5 times.

13.06V - 5.22V = 2.5
01.26V - 0.50V = 2.5

**Reply by pkkrusty** [reply]
There are usually 3 wires required for INA226.  V+, V-, and Vbus. Not attaching Vbus (normally also attached to V+) might be causing your issues.

**Reply by barbudor** [reply]
Yes, there are actually 4 wires to connect : V+, V-, VBus ad GND and all 4 must be connected
Of course the GND of the INA must also be connected to the GND of the ESP board

However there are 2 ways of using INA226 : High-side or Low-side

![image](https://github.com/arendst/Tasmota/assets/2996042/1579c56f-81ff-4773-be8d-63f0489c9d4a)

**Reply by tasmotari** [reply]
Thank you @barbudor @pkkrusty! Now all works)

**Reply by Quest24** [reply]
The problem with the factor 2.5 is due to a mixup with INA219 driver. This can be solved by disable this driver 14 with command  i2cdriver14 0. This has nothing to do with wiring!  (the driver is available even if the INA219 is not selected in the config)
